contracts/axelarnet-gateway/src/state.rs

```diff
@@ -51,288 +27,103 @@ pub enum Error {
-            message.cc_id.clone(),
-            message.clone(),
-        )
-        .unwrap();
-        let loaded_message =
-            super::may_load_received_msg(deps.as_ref().storage, &message.cc_id).unwrap();
-        assert_eq!(
-            Some(MessageWithStatus {
-                msg: message.clone(),
-                status: MessageStatus::Approved
-            }),
-            loaded_message
-        );
-
-        // Test saving mismatched message
-        let mismatched_message = Message {
-            cc_id: message.cc_id.clone(),
-            source_address: "different-address".parse().unwrap(),
-            ..message.clone()
-        };
-        let result = super::save_received_msg(
-            deps.as_mut().storage,
-            message.cc_id.clone(),
-            mismatched_message,
-        );
-        assert_eq!(result, Err(Error::MessageMismatch(message.cc_id)));
-    }
-
-    #[test]
-    fn set_msg_as_executed() {
-        let mut deps = mock_dependencies();
-        let message = create_test_message();
-
-        // Save a received message
-        super::save_received_msg(
-            deps.as_mut().storage,
-            message.cc_id.clone(),
-            message.clone(),
-        )
-        .unwrap();
-
-        // Test setting message as executed
-        let executed_message =
-            super::set_msg_as_executed(deps.as_mut().storage, message.cc_id.clone()).unwrap();
-        assert_eq!(message, executed_message);
-
-        // Verify the message status is now Executed
-        let loaded_message =
-            super::may_load_received_msg(deps.as_ref().storage, &message.cc_id).unwrap();
-        assert_eq!(
-            Some(MessageWithStatus {
-                msg: message.clone(),
-                status: MessageStatus::Executed
-            }),
-            loaded_message
-        );
-
-        // Test setting an already executed message
-        let result = super::set_msg_as_executed(deps.as_mut().storage, message.cc_id.clone());
-        assert_eq!(result, Err(Error::MessageAlreadyExecuted(message.cc_id)));
-
-        // Test setting a non-existent message
-        let non_existent_id = CrossChainId::new("non-existent", "id").unwrap();
-        let result = super::set_msg_as_executed(deps.as_mut().storage, non_existent_id.clone());
-        assert_eq!(result, Err(Error::MessageNotApproved(non_existent_id)));
-    }
+/// Update the status of a message to executed if it is in approved status, error otherwise.
+pub fn update_as_executed(
+    storage: &mut dyn Storage,
+    cc_id: &CrossChainId,
+    action: impl FnOnce(Message) -> Result<Message, Error>,
+) -> Result<Message, Error> {
+    let msg = match may_load_executable_msg(storage, cc_id)? {
+        None => Err(Error::MessageNotApproved(cc_id.clone())),
+        Some(ExecutableMessage::Executed(_)) => Err(Error::MessageAlreadyExecuted(cc_id.clone())),
+        Some(ExecutableMessage::Approved(msg)) => Ok(action(msg)?),
```

Comment:
```suggestion
        Some(ExecutableMessage::Approved(msg)) if validate(msg) => Ok(msg),
```
instead of an arbitrary action that could transform the message, only provide a validation function?